docs/design/fodc/watchdog-and-flight-recoder.md

```diff
@@ -0,0 +1,383 @@
+
+**Watchdog**: Periodically polls metrics from the BanyanDB container and forwards them to the Flight Recorder for buffering.
+
+**Flight Recorder**: Buffers metrics data using fixed-size circular buffers (RingBuffer) with in-memory storage, ensuring data persists even when the main BanyanDB process crashes.
+
+Together, these components capture and preserve metrics data to ensure that critical observability data is not lost during process crashes.
+
+### Responsibilities
+
+**Watchdog Component**
+- Polls metrics from BanyanDB at configurable intervals
+- Parses Prometheus text format metrics efficiently
+- Forwards collected metrics to Flight Recorder for buffering
+- Handles connection failures and retries gracefully
+- Monitors BanyanDB process health
+
+**Flight Recorder Component**
+- Maintains a fixed-size circular buffer (RingBuffer) per metric
+- Stores metrics in-memory to ensure fast access and persistence across process crashes
+- Manages buffer capacity and handles overflow scenarios using circular overwrite behavior
+- Ensures data integrity and prevents data loss during crashes
+
+### Component Interaction Flow
+
+```
+BanyanDB Metrics Endpoint
+         │
+         │ (HTTP GET /metrics)
+         ▼
+    Watchdog Component
+         │
+         │ (Poll at interval)
+         │
+         │ Parse Prometheus Format
+         │
+         │ Forward Metrics
+         ▼
+    Flight Recorder Component
+         │
+         │ Write to RingBuffer
+         │
+         │ (Per-metric buffers)
+         ▼
+    In-Memory Storage
+```
+
+## Component Design
+
+### 1. Watchdog Component
+
+**Purpose**: Periodically polls metrics from BanyanDB and forwards them to Flight Recorder
+
+#### Core Responsibilities
+
+- **Metrics Polling**: Polls metrics from BanyanDB metrics endpoint at configurable intervals
+- **Metrics Parsing**: Uses metrics package to parse Prometheus text format efficiently
+- **Error Handling**: Implements exponential backoff for transient failures
+- **Health Monitoring**: Tracks BanyanDB process health and reports status
+
+#### Core Types
+
+**`Watchdog`**
+```go
+type Watchdog struct {
+	client       *http.Client
+	url          string
+	interval     time.Duration
+}
+```
+
+#### Key Functions
+
+**`Start(ctx context.Context) error`**
+- Initializes the watchdog component
+- Starts polling loop with configurable interval
+- Sets up HTTP client with connection reuse
+- Begins periodic metrics collection
+
+**`Stop(ctx context.Context) error`**
+- Gracefully stops the polling loop
+- Closes HTTP connections
+- Ensures in-flight requests complete
+
+**`pollMetrics(ctx context.Context) ([]metrics.RawMetric, error)`**
+- Fetches raw metrics text from endpoint
+- Uses metrics package to parse Prometheus text format
+- Returns parsed metrics or error
+- Implements retry logic with exponential backoff
+
+#### Configuration Flags
```

Comment:
We need a flag: `--max-metrics-memory-usage-percentage` https://github.com/apache/skywalking-banyandb/blob/main/pkg/cgroups/memory.go#L36